pkg/providers/vsphere/virtualmachine/configspec.go

```diff
@@ -156,6 +156,33 @@ func CreateConfigSpec(
 		}
 	}
 
+	// The documentation for a ResourceAllocation object states that all (VM)
+	// fields must be set for certain APIs, including ImportVApp. A part of
+	// that API is used by the vpxd placement engine to construct fake VMs in
+	// order to support assignable hardware. Therefore:
+	//
+	// CPU allocation:
+	// - Ensure limit is set to -1 if there is a reservation.
+	//
+	// Memory allocation:
+	// - Ensure shares is not nil.
+	// - Ensure limit is set to -1 if there is a reservation.
+	if a := configSpec.CpuAllocation; a != nil {
```

Comment:
Does it make more sense to have these as a final `} else {` clause in the CPU or memory blocks above, since we're otherwise kind of spreading our defaulting logic out

Reply:
This way it handles it no matter how they started. That was intentional.